packages/jest-types/src/Circus.ts

```diff
@@ -31,9 +31,17 @@ export type Hook = {
   timeout: number | undefined | null;
 };
 
-export type EventHandler = (event: Event, state: State) => void;
+export type EventHandler = (event: Event, state: State) => void | Promise<void>;
 
-export type Event =
+export type Event = SyncEvent | AsyncEvent;
```

Comment:
why not make all events async?

Reply:
That handful of sync events have a very specific context of happening:

1. `error` - called either inside `NodeJS.UncaughtExceptionListener` or ` NodeJS.UnhandledRejectionListener`. Mind that both signatures are expected to return `void`, not a `Promise<void>`:

![globalErrorHandlers](https://user-images.githubusercontent.com/1962469/72253263-cec76100-3609-11ea-89ea-19430daeec4b.png)

2. `start_describe_definition` and `finish_describe_definition` - they are called right as you call the global function `describe(msg, fn)`:

![Screenshot from 2020-01-13 13-33-48](https://user-images.githubusercontent.com/1962469/72253446-3ed5e700-360a-11ea-95e1-a45595181e2a.png)

You see, to make those dispatches asynchronous, I should make the global `describe` asynchronous first, and that seems risky to me, because:

a. We change the signature of the global `describe` function, which is too pretentious, IMO.
b. We push our users into getting those nasty errors from the linters that pop whenever you call an asynchronous function in the global scope (outside of an async function):

![promise-then](https://user-images.githubusercontent.com/1962469/72253675-c15ea680-360a-11ea-8167-a01eac1537e8.png)
c. Last, but not least, errors thrown inside an asynchronous `describe()` won't have the same effect if they had been thrown in a synchronous one, right? Hence, it will break this logic, which is undesirable too:

```ts
  const asyncError = new ErrorWithStack(undefined, describeFn);
  if (blockFn === undefined) {
    asyncError.message = `Missing second argument. It must be a callback function.`;
    throw asyncError;
  }
  // ...
```

3. The same goes for `add_hook` (globals `beforeEach`, `afterEach`, ...) and `add_test` (globals `it`, `test`, `test.todo`, ...). I was not sure it was a good idea to change their signatures, so I decided to leave those events synchronous.

4. As for `setup` and `include_test_location_in_result`, it seems you are right. I am going to push a commit that makes them async and test it out.

Reply:
Note that we'll probably change these errors (`hook`, `describe`, `test`) to fail during test run, instead of during definition - see #9515. Only matters for 2.c and 3

I'm a bit afraid that an `async handleTestEvent` function won't behave correctly if we don't `await` all `dispatch`es. Thoughts?

Reply:
I think this should be a very marginal case when somebody wants to do something really asynchronous for `add_hook` and `add_test`. It's sad we can't do everything at once here, but I see no easy way in the spirit of this PR to accomplish the all-events-async goal. I really hope we can ignore this handful of events here. I have written the warnings in the docs already.